### ISSUE TITLE
Unable to pickle.dump `torch.dtype` objects when invoked by `python -m`

### ISSUE THREAD

**Issue by zasdfgbnm** [pytorch/pytorch#11172]
## Issue description

I have a very simple python package:
```
test_pickle/
├── __init__.py
└── test.py
```

Where the `__init__.py` has only one line:
```python
import torch
```
And the `test.py` just `pickle.dump` the `torch.float32`

```python
import pickle
import torch

if __name__ == '__main__':
    with open('test', 'wb') as f:
        pickle.dump(torch.float32, f)
    print('pickle success')
```

When run by `python test_pickle/test.py`, everything works normally and I successfully get a print `pickle success`. But when run by `python -m test_pickle.test`, I get an error message:

```
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/gaoxiang/tmp/test_pickle/test.py", line 6, in <module>
    pickle.dump(torch.float32, f)
TypeError: __reduce_ex__() takes exactly one argument (0 given)
```

I have no idea on what cause this problem, but it seems very strange...

## System Info
```
Collecting environment information...
PyTorch version: 0.5.0a0+a24163a
Is debug build: No
CUDA used to build PyTorch: 9.2.148

OS: Arch Linux
GCC version: (GCC) 8.2.0
CMake version: version 3.12.1

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: 
GPU 0: GeForce GTX 1080
GPU 1: GeForce GT 710

Nvidia driver version: 396.54
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

**Comment by zou3519** [comment]
When did you last build master?

**Comment by zasdfgbnm** [comment]
@zou3519 The problem is still there for the latest master:
```
commit 539579aa9a09850e374287eb58e6b1d773488e98
Author: Elias Ellison <eellison@fb.com>
Date:   2018-09-04
```

**Comment by vince62s** [comment]
I think it's Python 3.7 related. Got the same issue.

**Comment by zasdfgbnm** [comment]
Some progress here:

According to my experiment, this issue is because when invoked by `python -m`, python is trying to pickle the module `torch`. `torch` is a module and should not be pickled, usually, when a user try to pickle a module, python would raise an exception `TypeError: can't pickle module objects`, but unfortunately, the `__reduce_ex__` of module `torch` is mistakenly replaced by the `__reduce_ex__` of `_C._VariableFunctions`, which makes the error message very hard to understand.

https://github.com/pytorch/pytorch/pull/13884 fixes the issue that `__reduce_ex__` of module `torch` is replaced

**Comment by zasdfgbnm** [comment]
This seems to be fixed when I am testing on my `Python 3.7.3` now. Closing.
